File: src/gestorAplicación/externo/Banco.java
package gestorAplicación.externo;

import java.util.ArrayList;
import java.time.Instant;
import java.util.Date;
import gestorAplicación.interno.Categoria;
import gestorAplicación.interno.Corriente;
import gestorAplicación.interno.Cuenta;
import gestorAplicación.interno.Movimientos;
import gestorAplicación.interno.Suscripcion;
import gestorAplicación.interno.Usuario;
import gestorAplicación.interno.Ahorros;

public class Banco extends Estado {
	private static final long serialVersionUID = 2L;
	private static ArrayList<Banco> bancosTotales = new ArrayList<Banco>();
	public static final String nombreD = "Bancos";
	private ArrayList<Banco> bancosTotalesAux = new ArrayList<Banco>();
	private String nombre;
	private double comision;
	private Divisas divisa;
	private Estado estadoAsociado;
	private double prestamo;
	private boolean asociado = false;
	private ArrayList<String> dic = new ArrayList<String>();
	private ArrayList<Double> cionario = new ArrayList<Double>();
	
	//Funcionalidad Compra de Cartera
	private double desc_suscripcion = 0.2;
	private double desc_movimientos_porcentaje = 0.2;
	private int desc_movimientos_cantidad = 5;
	
	//Constructor
	
	public Banco(String nombre, double comision, Estado estado) {
		this.nombre = nombre;
		this.setEstadoAsociado(estado);
		this.comision = comision + this.getTasa_impuestos();
		bancosTotales.add(this);
	}
	
	public Banco(String nombre, double comision, Estado estado, double prestamo) {
		this.nombre = nombre;
		this.setEstadoAsociado(estado);
		this.comision = comision + this.getTasa_impuestos();
		this.setPrestamo(prestamo);
		bancosTotales.add(this);
	}
	
	public Banco() {}
	
	//Métodos
	
	public String mostrarBancosTotales() {
		if(Banco.bancosTotales.size() != 0) { 
			for (int i = 0; i < Banco.bancosTotales.size();) { 
				return(i + 1 + ". " + Banco.bancosTotales.get(i).getNombre()); 
				}
		}else { 
			return("No hay bancos en este momento, considere asociar bancos"); 
			}
		return ("");
	}
	
	public String mostrarDic(int k) {
		String clave= getDic().get(k);
		return clave;
	}
	
	public String mostrarDic() {
		
		return"";
	}
	
	//Funcionalidad de Suscripciones de Usuarios
	public Object comprobarSuscripción(Usuario usuario) {
		for(Usuario u : Usuario.getUsuariosTotales()) {
			if(usuario.getId() == u.getId()) {
				usuario.setLimiteCuentas(usuario.getSuscripcion().getLimiteCuentas());
				switch(usuario.getSuscripcion()) {
					case DIAMANTE:
						this.setComision(this.getComision() * 0.50);
						return ("Bienvenido " + usuario.getNombre() + ", eres un cliente " + usuario.getSuscripcion().name() + " de nuestro banco, "
								+ "por eso te cobramos " + (this.getComision() * 0.50) + " de comision");
					case ORO:
						this.setComision(this.getComision() * 0.65);
						return ("Bienvenido " + usuario.getNombre() + ", eres un cliente " + usuario.getSuscripcion().name() + " de nuestro banco, "
								+ "por eso te cobramos " + this.getComision() * 0.65 + " de comision");
					case PLATA:
						this.setComision(this.getComision() * 0.85);
						return ("Bienvenido " + usuario.getNombre() + ", eres un cliente " + usuario.getSuscripcion().name() + " de nuestro banco, "
								+ "por eso te cobramos " + this.getComision() * 0.85 + " de comision");
					case BRONCE:
						this.setComision(this.getComision());
						return ("Bienvenido " + usuario.getNombre() + ", eres un cliente " + usuario.getSuscripcion().name() + " de nuestro banco, "
								+ "por eso te cobramos " + this.getComision() + " de comision");	
					default:
						return ("No encontramos tu grado de suscripción, considera registrarte en nuestro banco.");
				}
			}
		} return ("No encontramos tu ID registrado en este banco, considera registrarte en nuestro banco.");
	}
	
	// Método funcionalidad Asesoramiento de inversiones
	public static Integer retornoPortafolio(int riesgo, double invertir, String plazo, Usuario user) {

		Ahorros cuentaAhorros = new Ahorros(user.getBancosAsociados().get(0), 1234, Divisas.COP, "Ahorros", 1.0);
		Corriente cuentaCorriente = new Corriente(user.getBancosAsociados().get(0), 1234, Divisas.COP, "Corriente");
		Usuario pepitoMaster1234 = new Usuario("pepito", "pepito", "pepito");
		cuentaAhorros.setTitular(pepitoMaster1234);
		cuentaCorriente.setTitular(pepitoMaster1234);

		double interes = Math.random() + riesgo;
		if (user.getCuentasAhorrosAsociadas().size() != 0
				&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() > invertir) {

			double x = user.getCuentasAhorrosAsociadas().get(0).getSaldo();
			double cobro = x * interes - x;
			Movimientos movimiento = new Movimientos(cuentaAhorros, user.getCuentasAhorrosAsociadas().get(0), cobro,
					Categoria.OTROS, Date.from(Instant.now()));

			if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() < invertir) {
				return 1;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() < invertir) {
				return 2;
			}

			else if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() > invertir) {
				return 3;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() > invertir) {
				return 4;
			}

		}

		else if (user.getCuentasAhorrosAsociadas().size() != 0
				&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() < invertir) {
			double x = user.getCuentasAhorrosAsociadas().get(0).getSaldo();
			double cobro = x * interes - x;
			Movimientos movimiento = new Movimientos(cuentaAhorros, user.getCuentasAhorrosAsociadas().get(0), cobro,
					Categoria.OTROS, Date.from(Instant.now()));

			if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() < invertir) {
				return 5;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() < invertir) {
				return 6;
			}

			else if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() > invertir) {
				return 7;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasAhorrosAsociadas().get(0).getSaldo() > invertir) {
				return 8;
			}
		}

		else if (user.getCuentasCorrienteAsociadas().size() != 0
				&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() > invertir) {

			double x = user.getCuentasCorrienteAsociadas().get(0).getDisponible();
			double cobro = x * interes - x;
			Movimientos movimiento = new Movimientos(cuentaCorriente, user.getCuentasCorrienteAsociadas().get(0), cobro,
					Categoria.OTROS, Date.from(Instant.now()));

			if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() < invertir) {
				return 1;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() < invertir) {
				return 2;
			}

			else if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() > invertir) {
				return 3;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() > invertir) {
				return 4;
			}

		}

		else {
			double x = user.getCuentasCorrienteAsociadas().get(0).getDisponible();
			double cobro = x * interes - x;
			Movimientos movimiento = new Movimientos(cuentaCorriente, user.getCuentasCorrienteAsociadas().get(0), cobro,
					Categoria.OTROS, Date.from(Instant.now()));

			if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() < invertir) {
				return 5;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() < invertir) {
				return 6;
			}

			else if (movimiento.impuestosMovimiento(interes)
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() > invertir) {
				return 7;
			}

			else if (movimiento.impuestosMovimiento(interes) == false
					&& user.getCuentasCorrienteAsociadas().get(0).getDisponible() > invertir) {
				return 8;
			}
		}

		return 0;
	}

	public static Banco bancoPortafolio(Usuario user) {
		Banco banco = null;
		if (user.getBancosAsociados().size() == 1) {
			banco = user.getBancosAsociados().get(0);
		} else {
			for (int i = 1; i <= user.getBancosAsociados().size(); i++) {
				if (user.getBancosAsociados().get(i - 1) != user.getBancosAsociados().get(i)) {
					banco = user.getBancosAsociados().get(i);
				} else {
					continue;
				}
			}
		}
		return banco;
	}

	public static double interesesPortafolio(Banco banco, Usuario user) {
		double interes = 0.0;

		for (int i = 0; i < user.getBancosAsociados().size(); i++) {
			if (user.getBancosAsociados().get(i) == banco) {
				interes = Math.round((interes + Math.random() + i) * 100.0) / 100.0;
			}
		}
		return interes;
	}
	
	public static Double verificarTasasdeInteres(Suscripcion suscripcion, Corriente cuenta){
		double interes = 0.0d;
		double descuento_movimientos = cuenta.getBanco().retornarDescuentosMovimientos(cuenta);
		double[] descuento_suscripcion = cuenta.getBanco().retornarDescuentosSuscripcion();
		double[] descuento_total = Banco.descuentoTotal(descuento_movimientos, descuento_suscripcion);
		switch(suscripcion) {
			case DIAMANTE:
				if (cuenta.getBanco().getInteres_bancario_corriente() >= descuento_total[3]) {
					interes = cuenta.getBanco().getInteres_bancario_corriente() - descuento_total[3];
				}
				else {
					interes = 0.0;
				}
			case ORO:
				if (cuenta.getBanco().getInteres_bancario_corriente() >= descuento_total[2]) {
					interes = cuenta.getBanco().getInteres_bancario_corriente() - descuento_total[2];
				}
				else {
					interes = 0.0;
				}
			case PLATA:
				if(cuenta.getBanco().getInteres_bancario_corriente() >= descuento_total[1]) {
					interes = cuenta.getBanco().getInteres_bancario_corriente() - descuento_total[1];
				}
				else {
					interes = 0.0;
				}
			case BRONCE:
				if(cuenta.getBanco().getInteres_bancario_corriente() >= descuento_total[0]) {
					interes = cuenta.getBanco().getInteres_bancario_corriente() - descuento_total[0];
				}
				else {
					interes = 0.0;
				}
		}
		return interes;
	}
	
	private double[] retornarDescuentosSuscripcion() {
		double[] descuento = new double[4];
		for(int i = 1; i < 5; i++) {
			descuento[i - 1] = this.desc_suscripcion * i;
		}
		return descuento;
	}
	
	//Método para recibir el descuento ocasionado por la cantidad de movimientos
	private double retornarDescuentosMovimientos(Corriente cuenta) {
		//Atributo que almacena todos los movimientos que están asociados del usuario con un Banco
		ArrayList<Movimientos> movimientosOriginariosconBanco = Movimientos.verificarMovimientosUsuario_Banco(cuenta.getTitular(), cuenta.getBanco());
		double descuento = Math.floorDiv(movimientosOriginariosconBanco.size(), this.desc_movimientos_cantidad) * this.desc_movimientos_porcentaje;
		return descuento;
	}
	
	public static ArrayList<Double> verificarTasasdeInteres(Usuario usuario, ArrayList<Corriente> cuentas){
		ArrayList<Double> tasasdeInteres = new ArrayList<Double>();
		Suscripcion suscripcion = usuario.getSuscripcion();
		for (Cuenta cuenta : cuentas) {
			Double interes = verificarTasasdeInteres(suscripcion, (Corriente) cuenta);
			tasasdeInteres.add(interes);
		}
		return tasasdeInteres;
	}
	
	public static double[] descuentoTotal(double movimientos, double[] suscripcion) {
		double[] descuento_total = suscripcion;
		for (double descuento : descuento_total) {
			descuento = descuento + movimientos;
		}
		return descuento_total;
	}
	
	//Gets
	public double getComision() {
		return comision;
	}
	
	public String getNombre() {
		return nombre;
	}

	public static ArrayList<Banco> getBancosTotales() {
		return bancosTotales;
	}
	
	public Estado getEstadoAsociado() {
		return estadoAsociado;
	}

	public double getPrestamo(){return prestamo;}
	
	public Divisas getDivisa() { return divisa; }
	
	public boolean isAsociado() {
		return asociado;
	}
	
	public ArrayList<String> getDic() {
		return dic;
	}
	
	public ArrayList<Double> getCionario() {
		return cionario;
	}

	//Sets
	public void setDivisa(Divisas divisa) { this.divisa = divisa; }
	
	public void setComision(double comision) {
		this.comision = comision;
	}
	
	public void setNombre(String nombre) {
		this.nombre = nombre;
	}

	public static void setBancosTotales(ArrayList<Banco> bancosTotales) {
		Banco.bancosTotales = bancosTotales;
	}

	public void setEstadoAsociado(Estado estadoAsociado) {
		this.estadoAsociado = estadoAsociado;
	}

	public void setPrestamo(Double prestamo){this.prestamo = prestamo;}
	
	public void setAsociado(boolean asociado) {
		this.asociado = asociado;
	}
	
	public void setDic(ArrayList<String> dic) {
		this.dic = dic;
	}
	
	public void setCionario(ArrayList<Double> cionario) {
		this.cionario = cionario;
	}
	
	public String toString() {
		return this.nombre;
	}

	public ArrayList<Banco> getBancosTotalesAux() {
		return bancosTotalesAux;
	}

	public void setBancosTotalesAux(ArrayList<Banco> bancosTotalesAux) {
		this.bancosTotalesAux = bancosTotalesAux;
	}
}

File: Python/src/gestorAplicación/externo/banco.py
import random
from datetime import datetime

from gestorAplicación.interno.suscripcion import Suscripcion
from gestorAplicación.interno.categoria import Categoria
from .estado import Estado

class Banco():
    _bancosTotales = []

    def __init__(self, nombre = "Banco de Colombia", comision = 0.3, estado = Estado(), prestamo = 200, **kwargs):
        #Revisar ingreso de atributo Estado, para que su defecto sea el primero como en Java
        self._nombre = nombre
        self._comision = comision
        self._estado = estado
        self._prestamo = prestamo

        self._cupo_base = 1000000
        self._multiplicador = 2
        self._desc_suscripcion = 0.2
        self._desc_movimientos_porcentaje = 0.2
        self._desc_movimientos_cantidad = 5

        #Atributos de instancia
        Banco._bancosTotales.append(self)
        self._id = len(Banco._bancosTotales)
        for key in kwargs:
            if key == "desc_suscripcion":
                self._desc_suscripcion = kwargs[key]
            if key == "desc_movimientos_porcentaje":
                self._desc_movimientos_porcentaje = kwargs[key]
            if key == "desc_movimientos_cantidad":
                self._desc_movimientos_cantidad = kwargs[key]
            if key == "divisa":
                self._divisa = kwargs[key]
            if key == "dic":
                self._dic = kwargs[key]
            if key == "cionario":
                self._cionario = kwargs[key]
            if key == "cupo_base":
                self._cupo_base
            if key == "multiplicador":
                self._multiplicador
            if key == "desc_suscripcion":
                self._desc_suscripcion
            if key == "desc_movimientos_porcentaje":
                self._desc_movimientos_porcentaje
            if key == "desc_movimientos_cantidad":
                self._desc_movimientos_cantidad

    #Métodos de la funcionalidad de cambio de divisa
    @staticmethod
    def cotizar_taza(user, existe_cambio, cadena, ahorros_posibles):
        from gestorAplicación.interno.movimientos import Movimientos
        imprimir = []
        for ahorro in ahorros_posibles:
            for banco in existe_cambio:
                indice = banco.get_dic().index(cadena)
                valor = banco.get_cionario()[indice]
                if ahorro.get_banco() == banco:
                    valor *= 0.98
                if banco.is_asociado():
                    valor *= 0.97
                cuota_manejo = Banco.divisa_suscripcion(user)
                cotizacion = Movimientos(banco, ahorro, valor, cuota_manejo)
                imprimir.append(cotizacion)

        return imprimir

    @staticmethod
    def cotizar_taza_aux(user, existe_cambio, cadena, cuentas_posibles):
        from gestorAplicación.interno.movimientos import Movimientos
        imprimir = []
        for conta in cuentas_posibles:
            for banco in existe_cambio:
                indice = banco.get_dic().index(cadena)
                valor = banco.get_cionario()[indice]
                if conta.get_banco() == banco:
                    valor *= 0.98
                if banco.is_asociado():
                    valor *= 0.97
                cuota_manejo = Banco.divisa_suscripcion(user)
                cotizacion = Movimientos(banco, conta, valor, cuota_manejo)
                imprimir.append(cotizacion)

        return imprimir



    # Métodos de la funcionalidad Asesoramiento de inversiones
    def retornoPortafolio(riesgo, invertir, plazo, user):
        from gestorAplicación.interno.usuario import Usuario
        from gestorAplicación.interno.movimientos import Movimientos
        interes = random.random() + riesgo

        if len(user.cuentasAhorrosAsociadas) != 0 and user.cuentasAhorrosAsociadas[0].saldo > invertir:
            cobro = user.cuentasAhorrosAsociadas[0].saldo / 32
            movimiento = Movimientos(
                user.cuentasAhorrosAsociadas[0],
                Usuario.getUsuariosTotales()[Usuario.hallarUsuarioImpuestosPortafolio()].cuentasAhorrosAsociadas[0],
                cobro, Categoria.OTROS, datetime.now()
            )
            if movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo < invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 1
            elif not movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo < invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 2
            elif movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo > invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 3
            elif not movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo > invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 4

        elif len(user.cuentasAhorrosAsociadas) != 0 and user.cuentasAhorrosAsociadas[0].saldo < invertir:
            cobro = user.cuentasAhorrosAsociadas[0].saldo / 32
            movimiento = Movimientos(
                user.cuentasAhorrosAsociadas[0],
                Usuario.getUsuariosTotales()[Usuario.hallarUsuarioImpuestosPortafolio()].cuentasAhorrosAsociadas[0],
                cobro, Categoria.OTROS, datetime.now()
            )
            if movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo < invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 5
            elif not movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo < invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 6
            elif movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo > invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 7
            elif not movimiento.impuestosMovimiento(interes) and user.cuentasAhorrosAsociadas[0].saldo > invertir:
                Movimientos.movimientosTotales.remove(movimiento)
                return 8

        elif len(user.cuentasCorrienteAsociadas) != 0 and user.cuentasCorrienteAsociadas[0].disponible > invertir:
            cobroCorriente = user.cuentasCorrienteAsociadas[0].disponible / 32
            movimientoCorriente = Movimientos(
                user.cuentasCorrienteAsociadas[0],
                Usuario.getUsuariosTotales()[Usuario.hallarUsuarioImpuestosPortafolio()].cuentasCorrienteAsociadas[0],
                cobroCorriente, Categoria.OTROS, datetime.now()
            )
            if movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible < invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 1
            elif not movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible < invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 2
            elif movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible > invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 3
            elif not movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible > invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 4

        else:
            cobroCorriente = user.cuentasCorrienteAsociadas[0].disponible / 32
            movimientoCorriente = Movimientos(
                user.cuentasCorrienteAsociadas[0],
                Usuario.getUsuariosTotales()[Usuario.hallarUsuarioImpuestosPortafolio()].cuentasCorrienteAsociadas[0],
                cobroCorriente, Categoria.OTROS, datetime.now()
            )
            if movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible < invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 5
            elif not movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible < invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 6
            elif movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible > invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 7
            elif not movimientoCorriente.impuestosMovimiento(interes) and user.cuentasCorrienteAsociadas[0].disponible > invertir:
                Movimientos.movimientosTotales.remove(movimientoCorriente)
                return 8

        return 0
    
    def bancoPortafolio(user):
        banco = None
        if len(user.bancosAsociados) == 1:
            banco = user.bancosAsociados[0]
        else:
            for i in range(1, len(user.bancosAsociados)):
                if user.bancosAsociados[i - 1] != user.bancosAsociados[i]:
                    banco = user.bancosAsociados[i]
                else:
                    continue
        return banco


    def interesesPortafolio(banco, user):
        interes = 0.0

        for i in range(len(user.bancosAsociados)):
            if user.bancosAsociados[i] == banco:
                interes = round((interes + random.random() + i), 2)
        return interes

    def decisionCupo(self, suscripcion):
        cupo = 0
        if suscripcion == Suscripcion.DIAMANTE:
            cupo = self._cupo_base * (self._multiplicador * 3)
        elif suscripcion == Suscripcion.ORO:
            cupo = self._cupo_base * (self._multiplicador * 2)
        elif suscripcion == Suscripcion.PLATA:
            cupo = self._cupo_base * (self._multiplicador)
        else:
            cupo = self._cupo_base
        return cupo
    
    def getNombre(self):
        return self._nombre
